0000-settlement-engines/0000-settlement-engines.md

```diff
@@ -0,0 +1,404 @@
+  - Positive amount indicates the connector is indebted to its peer (a _liability_ for the connector).
+  - Negative amount indicates the connector has sent a pre-payment to its peer.
+- **Accounts receivable**, the amount owed to the connector by its peer for packets the connector has fulfilled.
+  - Positive amount indicates its peer is indebted to the connector (an _asset_ to the connector).
+  - Negative amount indicates its peer has sent a pre-payment to the connector.
+
+Thus, a given connector's "accounts payable" balance should mirror its peer's "accounts receivable" balance. Likewise, a connector's "accounts receivable" balance should mirror its peer's "accounts payable" balance.
+
+## Settlement
+
+A **settlement** is the irrevocable discharge of a liability by providing something of value to the party owed.
+
+Settlements may occur on a **settlement system**, or medium for exchanging value. Some settlements may transfer funds on a **ledger**, or registry of account balances and/or transactions, which is a type of settlement system. (Although not all settlement systems are ledgers, here, the terms are sometimes used interchangeably.)
+
+Examples of settlement systems include:
+
+TODO Are these actually settlement systems?
+
+- Bank clearing houses
+- Credit card processors
+- Money transfer services
+- Cryptocurrencies, blockchains, and distributed ledgers
+- Payment channels and layer 2 networks
+- Cash or physical delivery of assets
+
+## Double-entry bookkeeping
+
+TODO Evan: Is it correct to refer to it as "credit" if you were prefunding or settling things right away? (It might be, but then it would be worth clarifying the different uses of credit -- IOUs, "credits vs debits", and this usage)
+
+TODO Explain clearing vs settlement?
+
+Together, a settlement engine and an accounting system interface with one another to perform double-entry bookkeeping. To ensure accurate, balanced double-entry bookkeeping, settlement engine and accounting system implementations MUST enforce several invariants.
+
+### Accounting system invariants
+
+#### Account for outgoing settlements
+
+The accounting system is responsible for triggering outgoing settlements. For example, when the accounts payable reaches a particular threshold, the accounting system SHOULD trigger a settlement to reduce the amount owed to the peer to a predefined, lesser amount in order to be able to continue transacting with the peer.
+
+If the accounting system opts to trigger a settlement:
+
+1. The accounting system MUST debit the accounts payable, subtracting the amount of the settlement.
+2. Then, the accounting system MUST send a request to the settlement engine to settle the amount.
+
+If the settlement engine responds that it only queued a partial amount for settlement (due to lesser precision), the accounting system MUST credit back the accounts payable, adding the leftover amount.
```

Comment:
Since we're going for eventual consistency, the engine will just respond with an "OK". The accounting system should not need to do anything with the response.

Reply:
Updated in the latest revision